providers/class-two-factor-webauthn.php

```diff
@@ -0,0 +1,642 @@
+					array(
+						'div'    => array(
+							'id'       => array(),
+							'class'    => array(),
+							'tabindex' => array(),
+						),
+						'ul'     => array(
+							'id'    => array(),
+							'class' => array(),
+						),
+						'li'     => array(
+							'id'    => array(),
+							'class' => array(),
+						),
+						'strong' => array(
+							'id'    => array(),
+							'class' => array(),
+						),
+						'small'  => array(
+							'id'    => array(),
+							'class' => array(),
+						),
+						'br'     => array(
+							'id'    => array(),
+							'class' => array(),
+						),
+						'em'     => array(
+							'id'    => array(),
+							'class' => array(),
+						),
+						'span'   => array(
+							'id'          => array(),
+							'class'       => array(),
+							'data-action' => array(),
+							'data-tested' => array(),
+							'tabindex'    => array(),
+						),
+						'a'      => array(
+							'id'          => array(),
+							'class'       => array(),
+							'data-action' => array(),
+							'tabindex'    => array(),
+							'href'        => array(),
+						),
+					)
+				);
+			}
+			?>
+		</ul>
+		<?php
+	}
+
+	/**
+	 * Registration Ajax Callback.
+	 */
+	public function ajax_register() {
+
+		check_ajax_referer( 'webauthn-register' );
+
+		if ( ! isset( $_REQUEST['payload'] ) ) {
+			// Error couldn't decode.
+			wp_send_json_error( new WP_Error( 'webauthn', __( 'Invalid request', 'two-factor' ) ) );
+		}
+
+		$credential = json_decode( wp_unslash( $_REQUEST['payload'] ) );
+
+		if ( JSON_ERROR_NONE !== json_last_error() ) {
+			// Error couldn't decode.
+			wp_send_json_error( new WP_Error( 'webauthn', esc_html( json_last_error_msg() ) ) );
+		}
+
+		if ( ! is_object( $credential ) ) {
+			// Contained some junk.
+			wp_send_json_error( new WP_Error( 'webauthn', __( 'Invalid credential', 'two-factor' ) ) );
+		}
+
+		// User id.
+		if ( isset( $_REQUEST['user_id'] ) ) {
+			$user_id = intval( wp_unslash( $_REQUEST['user_id'] ) );
+		} else {
+			wp_send_json_error( new WP_Error( 'webauthn', __( 'Invalid request data', 'two-factor' ) ) );
+		}
+		// Check permissions.
+		if ( ! current_user_can( 'edit_users' ) && get_current_user_id() !== $user_id ) {
+			wp_send_json_error( new WP_Error( 'webauthn', __( 'Not allowed to add key', 'two-factor' ) ) );
+		}
+
+		$keys = $this->key_store->get_keys( $user_id );
+
+		try {
+			$key = $this->webauthn->register( $credential, '' );
+
+			if ( false === $key ) {
+				wp_send_json_error( new WP_Error( 'webauthn', $this->webauthn->getLastError() ) );
+			}
+			/* translators: %s webauthn app id (domain) */
+			$key->label     = sprintf( esc_html__( 'New Device - %s', 'two-factor' ), $this->get_app_id() );
+			$key->md5id     = md5( implode( '', array_map( 'chr', $key->id ) ) );
+			$key->created   = time();
+			$key->last_used = false;
+			$key->tested    = false;
+
+			if ( false !== $this->key_store->find_key( $user_id, $key->md5id ) ) {
```

Comment:
This is probably not critical, but…

According to the [Registration Ceremony](https://www.w3.org/TR/webauthn-2/#sctn-registering-a-new-credential), Item 22, the credential must be unique site-wide:

> Check that the credentialId is not yet registered to any other user. If registration is requested for a credential that is already registered to a different user, the Relying Party SHOULD fail this registration ceremony, or it MAY decide to accept the registration, e.g. while deleting the older registration.